File: carkit/navigation/carkit_amcl/carkit_amcl/foxglove_waypoints.py
#!/usr/bin/env python3

from copy import deepcopy
from typing import List, Optional

import rclpy
from action_msgs.msg import GoalStatus
from geometry_msgs.msg import PoseStamped
from nav2_msgs.action import NavigateThroughPoses
from rclpy.action import ActionClient
from rclpy.node import Node
from rclpy.qos import DurabilityPolicy, QoSProfile, ReliabilityPolicy
from std_msgs.msg import String
from visualization_msgs.msg import Marker, MarkerArray


class FoxgloveWaypoints(Node):
    """Accumulate Foxglove pose clicks and submit them to Nav2."""

    def __init__(self) -> None:
        super().__init__('foxglove_waypoints')

        self.declare_parameter('goal_topic', '/foxglove/waypoints/goal')
        self.declare_parameter('command_topic', '/foxglove/waypoints/command')
        self.declare_parameter('marker_topic', '/foxglove/waypoints/markers')
        self.declare_parameter('status_topic', '/foxglove/waypoints/status')
        self.declare_parameter('action_name', '/navigate_through_poses')
        self.declare_parameter('default_frame', 'map')
        self.declare_parameter('maximum_poses', 50)

        status_qos = QoSProfile(depth=1)
        status_qos.reliability = ReliabilityPolicy.RELIABLE
        status_qos.durability = DurabilityPolicy.TRANSIENT_LOCAL

        self._pending: List[PoseStamped] = []
        self._active: List[PoseStamped] = []
        self._submitted: List[PoseStamped] = []
        self._sending = False
        self._goal_handle = None

        self._marker_pub = self.create_publisher(
            MarkerArray,
            str(self.get_parameter('marker_topic').value),
            status_qos,
        )
        self._status_pub = self.create_publisher(
            String,
            str(self.get_parameter('status_topic').value),
            status_qos,
        )
        self.create_subscription(
            PoseStamped,
            str(self.get_parameter('goal_topic').value),
            self._pose_callback,
            10,
        )
        self.create_subscription(
            String,
            str(self.get_parameter('command_topic').value),
            self._command_callback,
            10,
        )
        self._action_client = ActionClient(
            self,
            NavigateThroughPoses,
            str(self.get_parameter('action_name').value),
        )

        self._publish_markers()
        self._publish_status('Ready: publish poses to accumulate a route')

    def _pose_callback(self, msg: PoseStamped) -> None:
        maximum_poses = int(self.get_parameter('maximum_poses').value)
        if len(self._pending) >= maximum_poses:
            self._publish_status(f'Pose limit reached ({maximum_poses})')
            return

        pose = deepcopy(msg)
        if not pose.header.frame_id:
            pose.header.frame_id = str(self.get_parameter('default_frame').value)
        orientation = pose.pose.orientation
        if all(value == 0.0 for value in (
            orientation.x,
            orientation.y,
            orientation.z,
            orientation.w,
        )):
            pose.pose.orientation.w = 1.0

        self._pending.append(pose)
        self._publish_markers()
        self._publish_status(f'Accumulated {len(self._pending)} pose(s)')

    def _command_callback(self, msg: String) -> None:
        command = msg.data.strip().lower()
        if command in ('start', 'run', 'navigate'):
            self._start()
        elif command in ('clear', 'reset'):
            self._pending.clear()
            self._publish_markers()
            self._publish_status('Pending poses cleared')
        elif command in ('cancel', 'stop'):
            self._cancel()
        else:
            self._publish_status(
                f'Unknown command "{msg.data}"; use start, clear, or cancel'
            )

    def _start(self) -> None:
        if self._sending or self._goal_handle is not None:
            self._publish_status('A route is already starting or running')
            return
        if not self._pending:
            self._publish_status('No poses accumulated')
            return
        if not self._action_client.wait_for_server(timeout_sec=0.0):
            self._publish_status('Nav2 /navigate_through_poses is unavailable')
            return

        self._submitted = self._pending
        self._pending = []
        goal = NavigateThroughPoses.Goal()
        goal.poses = self._submitted

        self._sending = True
        self._publish_markers()
        self._publish_status(f'Starting route with {len(goal.poses)} pose(s)')
        future = self._action_client.send_goal_async(goal)
        future.add_done_callback(self._goal_response_callback)

    def _goal_response_callback(self, future) -> None:
        self._sending = False
        try:
            goal_handle = future.result()
        except Exception as exc:  # rclpy action transport failure
            self._pending = self._submitted + self._pending
            self._submitted = []
            self._publish_markers()
            self._publish_status(f'Failed to send route: {exc}')
            return

        if not goal_handle.accepted:
            self._pending = self._submitted + self._pending
            self._submitted = []
            self._publish_markers()
            self._publish_status('Nav2 rejected the route')
            return

        self._goal_handle = goal_handle
        self._active = self._submitted
        self._submitted = []
        self._publish_markers()
        self._publish_status(f'Navigating through {len(self._active)} pose(s)')
        result_future = goal_handle.get_result_async()
        result_future.add_done_callback(self._result_callback)

    def _result_callback(self, future) -> None:
        try:
            status = future.result().status
            status_text = {
                GoalStatus.STATUS_SUCCEEDED: 'Route completed',
                GoalStatus.STATUS_CANCELED: 'Route canceled',
                GoalStatus.STATUS_ABORTED: 'Route aborted by Nav2',
            }.get(status, f'Route ended with status {status}')
        except Exception as exc:  # rclpy action transport failure
            status_text = f'Failed to receive route result: {exc}'

        self._goal_handle = None
        self._active = []
        self._publish_markers()
        self._publish_status(status_text)

    def _cancel(self) -> None:
        if self._goal_handle is None:
            self._publish_status('No active route to cancel')
            return
        self._publish_status('Canceling active route')
        self._goal_handle.cancel_goal_async()

    def _publish_status(self, text: str) -> None:
        msg = String()
        msg.data = text
        self._status_pub.publish(msg)
        self.get_logger().info(text)

    def _publish_markers(self) -> None:
        marker_array = MarkerArray()
        clear = Marker()
        clear.action = Marker.DELETEALL
        marker_array.markers.append(clear)

        marker_id = 0
        for state, poses, color in (
            ('active', self._active, (0.1, 0.55, 1.0)),
            ('pending', self._pending, (0.1, 1.0, 0.35)),
        ):
            for index, pose in enumerate(poses, start=1):
                arrow = Marker()
                arrow.header = pose.header
                arrow.ns = f'foxglove_waypoints_{state}'
                arrow.id = marker_id
                marker_id += 1
                arrow.type = Marker.ARROW
                arrow.action = Marker.ADD
                arrow.pose = deepcopy(pose.pose)
                arrow.scale.x = 0.65
                arrow.scale.y = 0.14
                arrow.scale.z = 0.14
                arrow.color.r, arrow.color.g, arrow.color.b = color
                arrow.color.a = 1.0
                marker_array.markers.append(arrow)

                label = Marker()
                label.header = pose.header
                label.ns = f'foxglove_waypoints_{state}_labels'
                label.id = marker_id
                marker_id += 1
                label.type = Marker.TEXT_VIEW_FACING
                label.action = Marker.ADD
                label.pose = deepcopy(pose.pose)
                label.pose.position.z += 0.45
                label.scale.z = 0.35
                label.color.r, label.color.g, label.color.b = color
                label.color.a = 1.0
                label.text = str(index)
                marker_array.markers.append(label)

        self._marker_pub.publish(marker_array)


def main(args: Optional[List[str]] = None) -> None:
    rclpy.init(args=args)
    node = FoxgloveWaypoints()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: carkit/navigation/carkit_amcl/src/foxglove_waypoints.cpp
// Copyright 2026 University of Delaware
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <algorithm>
#include <cctype>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "action_msgs/msg/goal_status.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "nav2_msgs/action/navigate_through_poses.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "std_msgs/msg/string.hpp"
#include "visualization_msgs/msg/marker.hpp"
#include "visualization_msgs/msg/marker_array.hpp"

// CARKit learning annotation: implements the behavior described by this file's package and module.
class FoxgloveWaypoints : public rclcpp::Node
{
public:
  using Navigate = nav2_msgs::action::NavigateThroughPoses;
  using GoalHandle = rclcpp_action::ClientGoalHandle<Navigate>;

  FoxgloveWaypoints()
  : Node("foxglove_waypoints")
  {
    const auto goal_topic = declare_parameter<std::string>(
      "goal_topic", "/foxglove/waypoints/goal");
    const auto command_topic = declare_parameter<std::string>(
      "command_topic", "/foxglove/waypoints/command");
    const auto marker_topic = declare_parameter<std::string>(
      "marker_topic", "/foxglove/waypoints/markers");
    const auto status_topic = declare_parameter<std::string>(
      "status_topic", "/foxglove/waypoints/status");
    const auto action_name = declare_parameter<std::string>(
      "action_name", "/navigate_through_poses");
    default_frame_ = declare_parameter<std::string>("default_frame", "map");
    maximum_poses_ = std::max<int64_t>(1, declare_parameter("maximum_poses", 50));

    auto status_qos = rclcpp::QoS(1).reliable().transient_local();
    marker_publisher_ = create_publisher<visualization_msgs::msg::MarkerArray>(
      marker_topic, status_qos);
    status_publisher_ = create_publisher<std_msgs::msg::String>(status_topic, status_qos);
    pose_subscription_ = create_subscription<geometry_msgs::msg::PoseStamped>(
      goal_topic, 10,
      [this](geometry_msgs::msg::PoseStamped::ConstSharedPtr message) {on_pose(*message);});
    command_subscription_ = create_subscription<std_msgs::msg::String>(
      command_topic, 10,
      [this](std_msgs::msg::String::ConstSharedPtr message) {on_command(message->data);});
    action_client_ = rclcpp_action::create_client<Navigate>(this, action_name);
    publish_markers();
    publish_status("Ready: publish poses to accumulate a route");
  }

private:
  static std::string lower(std::string value)
  {
    std::transform(
      value.begin(), value.end(), value.begin(), [](unsigned char character) {
        return static_cast<char>(std::tolower(character));
      });
    return value;
  }

  void on_pose(const geometry_msgs::msg::PoseStamped & message)
  {
    if (pending_.size() >= static_cast<std::size_t>(maximum_poses_)) {
      publish_status("Pose limit reached (" + std::to_string(maximum_poses_) + ")");
      return;
    }
    auto pose = message;
    if (pose.header.frame_id.empty()) {
      pose.header.frame_id = default_frame_;
    }
    const auto & orientation = pose.pose.orientation;
    if (orientation.x == 0.0 && orientation.y == 0.0 &&
      orientation.z == 0.0 && orientation.w == 0.0)
    {
      pose.pose.orientation.w = 1.0;
    }
    pending_.push_back(std::move(pose));
    publish_markers();
    publish_status("Accumulated " + std::to_string(pending_.size()) + " pose(s)");
  }

  void on_command(const std::string & raw_command)
  {
    const auto command = lower(raw_command);
    if (command == "start" || command == "run" || command == "navigate") {
      start();
    } else if (command == "clear" || command == "reset") {
      pending_.clear();
      publish_markers();
      publish_status("Pending poses cleared");
    } else if (command == "cancel" || command == "stop") {
      cancel();
    } else {
      publish_status("Unknown command \"" + raw_command + "\"; use start, clear, or cancel");
    }
  }

  void start()
  {
    if (sending_ || active_goal_) {
      publish_status("A route is already starting or running");
      return;
    }
    if (pending_.empty()) {
      publish_status("No poses accumulated");
      return;
    }
    if (!action_client_->action_server_is_ready()) {
      publish_status("Nav2 /navigate_through_poses is unavailable");
      return;
    }

    submitted_ = std::move(pending_);
    pending_.clear();
    Navigate::Goal goal;
    goal.poses = submitted_;
    sending_ = true;
    publish_markers();
    publish_status("Starting route with " + std::to_string(goal.poses.size()) + " pose(s)");

    rclcpp_action::Client<Navigate>::SendGoalOptions options;
    options.goal_response_callback = [this](GoalHandle::SharedPtr goal_handle) {
        sending_ = false;
        if (!goal_handle) {
          pending_.insert(pending_.begin(), submitted_.begin(), submitted_.end());
          submitted_.clear();
          publish_markers();
          publish_status("Nav2 rejected the route");
          return;
        }
        active_goal_ = goal_handle;
        active_ = std::move(submitted_);
        submitted_.clear();
        publish_markers();
        publish_status("Navigating through " + std::to_string(active_.size()) + " pose(s)");
      };
    options.result_callback = [this](const GoalHandle::WrappedResult & result) {
        std::string status;
        switch (result.code) {
          case rclcpp_action::ResultCode::SUCCEEDED:
            status = "Route completed";
            break;
          case rclcpp_action::ResultCode::CANCELED:
            status = "Route canceled";
            break;
          case rclcpp_action::ResultCode::ABORTED:
            status = "Route aborted by Nav2";
            break;
          default:
            status = "Route ended with an unknown status";
            break;
        }
        active_goal_.reset();
        active_.clear();
        publish_markers();
        publish_status(status);
      };
    action_client_->async_send_goal(goal, options);
  }

  void cancel()
  {
    if (!active_goal_) {
      publish_status("No active route to cancel");
      return;
    }
    publish_status("Canceling active route");
    action_client_->async_cancel_goal(active_goal_);
  }

  void publish_status(const std::string & text)
  {
    std_msgs::msg::String message;
    message.data = text;
    status_publisher_->publish(message);
    RCLCPP_INFO(get_logger(), "%s", text.c_str());
  }

  void append_markers(
    visualization_msgs::msg::MarkerArray & output,
    const std::vector<geometry_msgs::msg::PoseStamped> & poses,
    const std::string & state, float red, float green, float blue, int & marker_id) const
  {
    int index = 1;
    for (const auto & pose : poses) {
      visualization_msgs::msg::Marker arrow;
      arrow.header = pose.header;
      arrow.ns = "foxglove_waypoints_" + state;
      arrow.id = marker_id++;
      arrow.type = visualization_msgs::msg::Marker::ARROW;
      arrow.action = visualization_msgs::msg::Marker::ADD;
      arrow.pose = pose.pose;
      arrow.scale.x = 0.65;
      arrow.scale.y = 0.14;
      arrow.scale.z = 0.14;
      arrow.color.r = red;
      arrow.color.g = green;
      arrow.color.b = blue;
      arrow.color.a = 1.0;
      output.markers.push_back(arrow);

      auto label = arrow;
      label.ns += "_labels";
      label.id = marker_id++;
      label.type = visualization_msgs::msg::Marker::TEXT_VIEW_FACING;
      label.pose.position.z += 0.45;
      label.scale.x = 0.0;
      label.scale.y = 0.0;
      label.scale.z = 0.35;
      label.text = std::to_string(index++);
      output.markers.push_back(label);
    }
  }

  void publish_markers()
  {
    visualization_msgs::msg::MarkerArray output;
    visualization_msgs::msg::Marker clear;
    clear.action = visualization_msgs::msg::Marker::DELETEALL;
    output.markers.push_back(clear);
    int marker_id = 0;
    append_markers(output, active_, "active", 0.1F, 0.55F, 1.0F, marker_id);
    append_markers(output, pending_, "pending", 0.1F, 1.0F, 0.35F, marker_id);
    marker_publisher_->publish(output);
  }

  std::string default_frame_;
  int64_t maximum_poses_{50};
  bool sending_{false};
  std::vector<geometry_msgs::msg::PoseStamped> pending_;
  std::vector<geometry_msgs::msg::PoseStamped> active_;
  std::vector<geometry_msgs::msg::PoseStamped> submitted_;
  GoalHandle::SharedPtr active_goal_;
  rclcpp_action::Client<Navigate>::SharedPtr action_client_;
  rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr marker_publisher_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr status_publisher_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr pose_subscription_;
  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr command_subscription_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<FoxgloveWaypoints>());
  rclcpp::shutdown();
  return 0;
}
